File: DevSharp/Domain/IAggregateClass.cs
using System.Collections.Generic;
using DevFSharp;

namespace DevSharp.Domain
{
    public interface IAggregateClass
    {
        IEnumerable<Validations.ValidationResult> ValidateCommand(object command);
        IEnumerable<object> ExecuteCommand(object currentState, object command);
        object ApplyEvent(object currentState, object @event);
    }
}

File: DevSharp/Domain/IAggregateClass.fs
namespace DevSharp.Domain

open FSharp.Core
open DevSharp.Messaging
open DevSharp.Validations

type CommandType = obj
type EventType   = obj
type StateType   = obj


type IAggregateClass =

    abstract member init: StateType
    abstract member isStateless: bool

    abstract member requiresRequest: bool

    abstract member validate: CommandType -> Request -> ValidationResult

    abstract member act: CommandType -> StateType -> Request -> EventType seq

    abstract member apply: EventType -> StateType -> Request -> StateType
